framework/src/modules/chain/submodules/delegates.js

```diff
@@ -587,35 +589,55 @@ __private.forge = function(cb) {
 				`Broadhash consensus before forging a block: ${modules.peers.getLastConsensus()} %`
 			);
 
-			return modules.blocks.process.generateBlock(
-				delegateKeypair,
-				slots.getSlotTime(currentSlot),
-				blockGenerationErr => {
-					if (blockGenerationErr) {
-						library.logger.error(
-							'Failed to generate block within delegate slot',
-							blockGenerationErr
-						);
+			// If last block slot is way back than one block
+			if (lastBlockSlot < currentSlot - 1) {
+				library.logger.info(
+					`Waiting for ${
+						library.config.forging.waitThreshold
+					} seconds for the last block before forging`
+				);
+				library.logger.debug('Slot information', {
+					currentSlot,
+					lastBlockSlot,
+				});
+				return setTimeout(() => {
+					__private.generateBlock(delegateKeypair, currentSlot, cb);
```

Comment:
I don't think we should use setTimeout here for two reasons:
1- It's possible that we receive the block from the previous slot before this setTimeout is executed and if the lastBlockSlot was the last slot of the round, we will call __private.generateBlock with invalid keypair slot based because the round is changed.
2- __private.forge is executed every second, therefore we don't need to setTimeout anyway.

Reply:
Not convinced with point 1 as even though we receive block meanwhile the currentSlot and keyapir for that slot does not change. 

Yes for the second point, yes we can skip in case we want to wait.

Reply:
If it's the end of the round and the last block is not the most recent one, then this line will have different result: 
```
const round = slots.calcRound(lastBlock.height + 1);
```
Therefore, the round parameter passed to this function will be different:
```
__private.getDelegateKeypairForCurrentSlot(currentSlot, round)
```
Therefore, the response returned from this function will be invalid because it's from a previous round.